### ISSUE TITLE
Warnings in templates should be triggered based on flags of instantiating module, not instantiated module

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#60089]
Given a.cpp:

```cpp
export module a;

export template<typename T>
void a(T y) {
	unsigned x = 0;
	x = x + y;
}
```

compiled with

`clang++ -std=c++20 -Werror -Wshorten-64-to-32 -x c++-module a.cpp --precompile -o a.pcm`

and b.cpp:

```cpp
export module b;

import a;

void b() {
	a(static_cast<unsigned long>(1));
}
```

compiled with

`clang++ -std=c++20 -Wno-error -Wno-shorten-64-to-32 b.cpp -fprebuilt-module-path=. a.pcm -o /dev/null`

fails with

```shell
In file included from b.cpp:3:
a.cpp:6:8: error: implicit conversion loses integer precision: 'unsigned long' to 'unsigned int' [-Werror,-Wshorten-64-to-32]
        x = x + y;
          ~ ~~^~~
b.cpp:6:2: note: in instantiation of function template specialization 'a<unsigned long>' requested here
        a(static_cast<unsigned long>(1));
        ^
1 error generated.
```

So to summarize, I compile b.cpp with "warnings not as errors, and don't warn me about this thing anyway", and clang warns me about it and makes it an error, because I happened to compile a.cpp with a different set of warning flags.

**Comment by ChuanqiXu9** [comment]
Oh, great. The suggestion makes  sense.

BTW (not specified to this issue), currently we mainly focus on fixing bugs and implementing new features. The priority of the diagnostic related problems is relatively low. For example, in the message you quoted, it said "a.cpp is included from b.cpp". But it is not true. There are many other similar problems. Such diagnostic-related issues are pretty helpful, although they may not be addressed quickly.
